### PR TITLE
Instruction flags constraints, cleanup

### PR DESCRIPTION
**Pull request by arasuarun**
working, but needs a TON of cleanup

### CODE REVIEW

File: jolt-core/src/jolt/vm/mod.rs [hunk 1]
```diff
@@ -505,8 +507,8 @@ pub trait Jolt<F: PrimeField, G: CurveGroup<ScalarField = F>, const C: usize, co
         let inputs: R1CSInputs<F> = R1CSInputs::new(
             padded_trace_len,
             bytecode_a,
-            bytecode_v,
-            packed_flags,
+            bytecode_v.clone(),
```
Comment (sragss) [inline jolt-core/src/jolt/vm/mod.rs:510]: Looks like these are cloned just to be able to print afterwards

**Comment by arasuarun** [comment]
Ready for review. Now, both the instruction flags and circuit flags are sent to the R1CS and are checked against the packed flags value `op_flags_packed` which replaces the opcode in the program code. (We don't actually need the opcode in the circuit.)

File: jolt-core/src/host/mod.rs [hunk 2]
```diff
@@ -121,11 +122,24 @@ impl Program {
...
             .flat_map(|row| {
-                row.instruction
+                let mut bitvector = row.instruction
                     .to_circuit_flags()
                     .iter()
                     .map(|&flag| flag.into())
-                    .collect::<Vec<F>>()
+                    .collect::<Vec<F>>();
+
+                let N_FLAGS = bitvector.len();
```
Comment (moodlezoup) [inline jolt-core/src/host/mod.rs:131]: Let's use `NUM_CIRCUIT_FLAGS` from `rv_trace.rs`

File: jolt-core/src/r1cs/constraints.rs [hunk 3]
```diff
@@ -7,7 +7,7 @@ use common::constants::{RAM_START_ADDRESS, RAM_WITNESS_OFFSET};
 
 /* Compiler Variables */
 const C: usize = 4; 
-const N_FLAGS: usize = 17; 
+const N_FLAGS: usize = 17 + 19;
```
Comment (moodlezoup) [inline jolt-core/src/r1cs/constraints.rs:10]: `NUM_CIRCUIT_FLAGS + RV32I::COUNT`

File: jolt-core/src/host/mod.rs [hunk 4]
```diff
@@ -121,11 +122,24 @@ impl Program {
...
+
+                // instruction flag
+                if let Ok(jolt_instruction) = InstructionSet::try_from(&row.instruction) {
+                    let instruction_index = InstructionSet::enum_index(&jolt_instruction);
+                    // bitvector <<= instruction_index;
+                    // bitvector |= 1;
+                    // bitvector <<= InstructionSet::COUNT - instruction_index - 1;
+                    bitvector[N_FLAGS + instruction_index] = F::from(1 as u64);
+                } 
+                bitvector
```
Comment (moodlezoup) [inline jolt-core/src/host/mod.rs:142]: We already have the instruction flag polynomials from `instruction_lookups`, shouldn't we be feeding those to the r1cs instead of recreating them here?

**Review by moodlezoup — CHANGES_REQUESTED** [review]

File: common/src/rv_trace.rs [hunk 5]
```diff
@@ -235,6 +235,18 @@ pub struct ELFInstruction {
...
+            address: 0,
+            opcode,
+            raw: 0,
+            rs1: None,
+            rs2: None,
+            rd: None,
+            imm: None,
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (arasuarun) [reply common/src/rv_trace.rs:249]: You're right. Removed it.

File: jolt-core/src/host/mod.rs (hunk 4 above)
Comment (arasuarun) [reply jolt-core/src/host/mod.rs:142]: Yeah this isn't ideal. I don't think the flags (either circuit or instr) should be in this `trace` function. It would be ideal to just use the opcode and get all the circuit and instruction flags in `mod.rs`. I tried doing something like this but I wasn't able to call `bitflags()` in bytecode.rs because it is expecting an ELFInstruction while we only have access to an RV32I enum in the `prove_r1cs` function. 

I might just be missing something, though. We could try to do a full refactor (no flags in trace) or a partial one for now and do the full later. What do you think?

Comment (moodlezoup) [reply jolt-core/src/host/mod.rs]: no what I mean is, we still need to compute `row.instruction.to_circuit_flags()` here in `trace`. But we don't need these instruction flags in the bitvector, because we already have them in the form of `InstructionPolynomials.instruction_flag_polys`
We should be passing those `instruction_flag_polys` into `prove_r1cs`, and using those in the circuit. As it is right now, we're recommitting to these instruction flag bits

Comment (arasuarun) [reply jolt-core/src/host/mod.rs]: Ah right. Yeah that should be the way to do it. Lemme try adding that now.

File: jolt-core/src/host/mod.rs
Comment (moodlezoup) [inline jolt-core/src/host/mod.rs:1]: nit: can we revert the changes to this file?

**Review by moodlezoup — APPROVED** [review]
one nit but otherwise lgtm

File: jolt-core/src/host/mod.rs
Comment (arasuarun) [reply jolt-core/src/host/mod.rs:1]: Done
